src/lib/Generator.class.ts

```diff
@@ -46,57 +47,59 @@ class Generator extends EventEmitter {
 
     this.engine = getEngine(this.endpoint)
 
-    stage.iterator.on('end', count => {
-      this.iterationsIncoming = count
-        for (const $this of this.$thisList) {
-          this.run($this, this.$thisList.length)
-      }
+    stage.iterator.on('end', _count => {
+      this.flush();
     })
+  }
 
+  public run($this: NamedNode, batchSize?: number): void {
+    this.$thisList.push($this)
+    this.iterationsIncoming++;
+    if (this.$thisList.length >= (batchSize ?? this.batchSize)) {
+      this.runBatch(this.$thisList);
+      this.$thisList = [];
+    }
   }
 
   private get batchSize(): number {
     return this.stage.configuration.generator[this.index].batchSize ?? DEFAULT_BATCH_SIZE
   }
 
-
-  public run($this?: NamedNode, batchSize?: number): void {
-    if ($this !== undefined) this.$thisList.push($this)
```

Comment:
Here the URIs are added to the same `this.$thisList` property again, causing duplicates.over.